### DISCUSSION TITLE
Get the client IP address in CustomCertificateValidator

### DISCUSSION THREAD

**Discussion by onidio** [CoreWCF/CoreWCF#1416] [Q&A]
We check the validity of the certificate with a custom validator:
credentials.ClientCertificate.Authentication.CustomCertificateValidator = new CustomCertificateValidator();
How to obtain the client's ip address in the Validate procedure so that it can be logged

        public override void Validate(X509Certificate2 certificate)
        {
             var clienIP=??;
        }

**Comment by mconnew** [comment]
There is a way, but it's a little convoluted. NetTcp using cert auth uses the binding element `SslStreamSecurityBindingElement`, which derives from `StreamUpgradeBindingElement`. You can create your own custom implementation of `StreamUpgradeBindingElement` which wraps `SslStreamSecurityBindingElement`. The method `BuildServerStreamUpgradeProvider` returns a `StreamUpgradeProvider`. You would call into the wrapped Ssl implementation and wrap the returned `StreamUpgradeProvider`. In your stream upgrade provider `CreateUpgradeAcceptor` method, you would call the same method on the wrapped upgrade provider, wrap that in your own acceptor and return that. There's a new (to CoreWCF compared with WCF) property on `StreamUpgradeAccepter`:
```c#
        public virtual IFeatureCollection Features { get; }= new FeatureCollection();
```
As part of establishing the connection, we add a feature to the feature collection of type `CoreWCF.Channels.Framing.FramingConnection`. In your stream upgrade acceptors implementation of `AcceptUpgradeAsync`, you would call the wrapped implementations method, but wrap it in a try/catch block. In the catch block, you can get the `FramingConnection` from the features collection. From there, there's this property:
```c#
public IPEndPoint RemoteEndpoint { get; }
```
So you can't get the remote IP address from the Validate method itself, but you can wrap the call which establishes an SslStream, which in turn calls your Validate method, which will throw an exception, which you can catch, and then fetch the remote IP address. My preference would be to follow through the on the other discussion to get things to work the same in CoreWCF as it does in WCF, but if you're in a hurry, this can get you a solution right now.

**Reply by onidio** [reply]
@mconnew
I will try to implement this;
But if the OnConnectedAsync(FramingConnection) solution already exists, I'd be interested in a quick code example

**Reply by onidio** [reply]
> There is a way, but it's a little convoluted. NetTcp using cert auth uses the binding element `SslStreamSecurityBindingElement`, which derives from `StreamUpgradeBindingElement`. You can create your own custom implementation of `StreamUpgradeBindingElement` which wraps `SslStreamSecurityBindingElement`. The method `BuildServerStreamUpgradeProvider` returns a `StreamUpgradeProvider`. You would call into the wrapped Ssl implementation and wrap the returned `StreamUpgradeProvider`. In your stream upgrade provider `CreateUpgradeAcceptor` method, you would call the same method on the wrapped upgrade provider, wrap that in your own acceptor and return that. There's a new (to CoreWCF compared with WCF) property on `StreamUpgradeAccepter`:
> 
> ```cs
>         public virtual IFeatureCollection Features { get; }= new FeatureCollection();
> ```
> 
> As part of establishing the connection, we add a feature to the feature collection of type `CoreWCF.Channels.Framing.FramingConnection`. In your stream upgrade acceptors implementation of `AcceptUpgradeAsync`, you would call the wrapped implementations method, but wrap it in a try/catch block. In the catch block, you can get the `FramingConnection` from the features collection. From there, there's this property:
> 
> ```cs
> public IPEndPoint RemoteEndpoint { get; }
> ```
> 
> So you can't get the remote IP address from the Validate method itself, but you can wrap the call which establishes an SslStream, which in turn calls your Validate method, which will throw an exception, which you can catch, and then fetch the remote IP address. My preference would be to follow through the on the other discussion to get things to work the same in CoreWCF as it does in WCF, but if you're in a hurry, this can get you a solution right now.

@mconnew
Thank you very much, it works. I just have some doubts about the implementation of OnAbort(), OnCloseAsync(CancellationToken token) and  OnOpenAsync(CancellationToken token)

```
internal class MyStreamUpgradeProvider : StreamUpgradeProvider
{
    private readonly StreamUpgradeProvider _Wrapped;

    public MyStreamUpgradeProvider(StreamUpgradeProvider streamUpgradeProvider)
    {
        _Wrapped = streamUpgradeProvider;
    }

    protected override void OnAbort()
    {
        _Wrapped.Abort();
    }

    protected override Task OnCloseAsync(CancellationToken token)
    {
        return _Wrapped.CloseAsync(token);
    }

    protected override Task OnOpenAsync(CancellationToken token)
    {
        return _Wrapped.OpenAsync(token);
    }

    public override StreamUpgradeAcceptor CreateUpgradeAcceptor()
    {
        return new MyStreamUpgradeAcceptor(_Wrapped.CreateUpgradeAcceptor());
        //var result = _Wrapped.CreateUpgradeAcceptor();
    }
}
```

**Reply by mconnew** [reply]
Calling the method without the On prefix on the wrapped implementation from an On method is the right way to delegate a call, but I think you also need to call the base implementation. The parent StreamUpgradeProvider transitions a state value when calling OpenAsync/CloseAsync/Abort to make sure you can't do things like call OpenAsync twice. Any real work happens in the On method. You should call base.OnXXXXX before calling the wrapped method. I'm not sure if StreamUpgradeProvider enforces this, but most classes with this pattern will throw if you don't do call the base method.
